library/core/src/hint.rs

```diff
@@ -219,6 +219,73 @@ pub fn spin_loop() {
 /// backend used. Programs cannot rely on `black_box` for *correctness* in any way.
 ///
 /// [`std::convert::identity`]: crate::convert::identity
+///
+/// # When is this needed?
+///
+/// The most common use case for `black_box` is to make sure that benchmarks actually run the code
+/// they contain, rather than being allowed to optimize parts away based on context. Observe the
+/// example:
+///
+/// ```
+/// fn contains(haystack: &[&str], needle: &str) -> bool {
+///     haystack.iter().any(|x| x == &needle)
+/// }
+///
+/// pub fn benchmark() {
+///     let haystack = vec!["abc", "def", "ghi", "jkl", "mno"];
+///     let needle = "ghi";
+///     for _ in 0..10 {
+///         contains(&haystack, needle);
+///     }
+/// }
+/// ```
+///
+/// The compiler _could_ theoretically make optimizations like the following:
+///
+/// - `needle` and `haystack` are always the same, move the call to `contains` outside the loop and
+///   delete the loop
+/// - Inline `contains`
+/// - `needle` and `haystack` have values known at compile time, `contains` is always true. Remove
+///   the call and replace with `true`
+/// - Nothing is done with the result of `contains`: delete this function call entirely
+/// - Benchmark now has no purpose: delete this function
+///
+/// It is not likely that all of the above happens, but the compiler is definitely able to make
+/// some optimizations that would result in a very inaccurate benmark. This is where `black_box`
+/// comes in:
+///
+/// ```
+/// use std::hint::black_box;
+///
+/// // Same `contains` function
+/// fn contains(haystack: &[&str], needle: &str) -> bool {
+///     haystack.iter().any(|x| x == &needle)
+/// }
+///
+/// // Adjust our benchmark
+/// pub fn benchmark() {
+///     let haystack = vec!["abc", "def", "ghi", "jkl", "mno"];
+///     let needle = "ghi";
+///     for _ in 0..10 {
+///         black_box(contains(black_box(&haystack), black_box(needle)));
+///     }
+/// }
+/// ```
+///
+/// This essentially tells the compiler to block optimizations across any calls to `black_box`. So,
+/// it now:
+///
+/// - Treats both arguments to `contains` as unpredictable: the body of `contains` cannot be
+///   optimized based on their values
+/// - Treats the call to `contains` and its result as volatile: the body of `benchmark` cannot
+///   optimize this away
+///
+/// This makes our benchmark much more realistic to how the function would be used in situ, where
+/// arguments are usually not known at compile time and the result is used in some way.
+///
+/// It is important to note that `black_box` cannot stop all optimizations across its call in all
```

Comment:
I think this might be understating a little - I know there was some discussion prior to stabilization/on the release blog post about emphasizing more of the lack of guarantees (cc @nagisa @saethlin).

I think at minimum, we shouldn't try to phrase it as "not stopping all optimizations", which makes it sound like some are guaranteed.

Reply:
Would something like this be better?

> # A word on guarantees 
>
> It is important to reiterate the above: `black_box` does _not_ guarantee with any certainty what optimizations may or may not happen. Variations such as different compiler backends, different targets,
or even just corner case code may mean it has no effects at all. The Rust team attempts to make sure this doesn't happen, but it cannot be guaratnteed: for that reason, `black_box` should _never_ be relied on for controlling behavior outside of test cases or benchmarks. 
>
> For the most performance-critical cases, there is no substitute for validating the generated assembly and using performance tools (e.g. `perf` or `callgrind`) to analyze your code. That being said, `black_box` is suitable for creating fairly accurate benchmarks in most cases, for most users.

I don't think I know enough about the specific context to know how accurate that is, so open to suggestions (didn't see too much on #64102)

Reply:
I would prefer changing the wording of the text you're proposing adding, as opposed to adding another section. Documentation here should emphasize everywhere that `black_box` having any effect at all on a program is entirely best-effort, it may do nothing at all, and therefore this is _only_ for benchmarks. The are _already_ people who think it is for writing cryptography code, and that is wrong.

Reply:
_In practice_ this function is pretty reliable, but it is absolutely NOT completely reliable, and even if it were, documenting what it actually accomplishes if it were reliable sounds like a nightmare.

Reply:
I've rearranged this somewhat and now that bit is at the beginning: https://github.com/rust-lang/rust/blob/79cf860f397aff83d283ac8de0ac77057ec59147/library/core/src/hint.rs#L225-L233